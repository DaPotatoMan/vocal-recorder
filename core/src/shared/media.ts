import { RecorderError } from './error'

export class StreamUtil {
  static async get(options?: MediaTrackConstraints) {
    const ctx: any = navigator.mediaDevices || navigator || {}
    const getMedia: MediaDevices['getUserMedia'] = (ctx.getUserMedia || ctx.getUserMedia
      || ctx.webkitGetUserMedia || ctx.mozGetUserMedia || ctx.msGetUserMedia)

    if (!getMedia) throw new RecorderError('NO_GETUSERMEDIA')

    const stream = await getMedia.call(ctx, { audio: options || true, video: false })
    return new MediaStream(stream)
  }

  static getSettings(stream: MediaStream) {
    return stream.getAudioTracks()[0].getSettings()
  }

  static isValid(stream: MediaStream) {
    const tracks = stream.getAudioTracks()
    return stream.active && tracks.length > 0 && tracks.some(e => e.enabled)
  }

  static dispose(stream: MediaStream) {
    stream.getTracks().forEach((entry) => {
      entry.stop()
      stream.removeTrack(entry)
    })

    return stream.active
  }
}

export function getAudioContext(options?: AudioContextOptions) {
  const ctx = globalThis
  const Ref = ctx.AudioContext || ctx.webkitAudioContext

  if (!Ref) throw new RecorderError('NO_AUDIO_CONTEXT')

  return new Ref(options)
}

export async function getAudioBuffer(buffer: ArrayBuffer, config: OfflineAudioContextOptions = {
  numberOfChannels: 1,
  sampleRate: 48000,
  length: 48000
}) {
  try {
    return new OfflineAudioContext(config).decodeAudioData(buffer)
  }
  catch {
    return getAudioContext({ sampleRate: config.sampleRate }).decodeAudioData(buffer)
  }
}

// Converts the Blob data to AudioBuffer
export async function getBlobAudioBuffer(blob: Blob) {
  return getAudioBuffer(await blob.arrayBuffer())
}
