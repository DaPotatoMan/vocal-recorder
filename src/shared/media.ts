import { RuntimeError } from './core'
import { getWindow } from './utils'

export class StreamUtil {
  static async get(options?: MediaTrackConstraints) {
    const ctx: any = navigator.mediaDevices || navigator || {}
    const getMedia: MediaDevices['getUserMedia'] = (ctx.getUserMedia || ctx.getUserMedia
      || ctx.webkitGetUserMedia || ctx.mozGetUserMedia || ctx.msGetUserMedia)

    if (!getMedia)
      throw new RuntimeError('GETUSERMEDIA_UNSUPPORTED')

    const stream = await getMedia.call(ctx, { audio: options || true, video: false })
    return new MediaStream(stream)
  }

  static getSettings(stream: MediaStream) {
    return stream.getAudioTracks()[0].getSettings()
  }

  static getCapabilities(stream: MediaStream) {
    return stream.getAudioTracks()[0].getCapabilities()
  }

  static isValid(stream: MediaStream) {
    const tracks = stream.getAudioTracks()
    return stream.active && tracks.length > 0 && tracks.some(e => e.enabled)
  }

  static dispose(stream: MediaStream) {
    for (const track of stream.getTracks()) {
      track.stop()
      stream.removeTrack(track)
    }

    stream.dispatchEvent(new Event('dispose'))
    return stream.active
  }

  /** Fires when `StreamUtil.dispose` is called on the given stream */
  static onDispose(stream: MediaStream, callback: EventListener) {
    stream.addEventListener('dispose', callback)
  }
}

export function getAudioContext(options?: AudioContextOptions) {
  try {
    const window = getWindow()
    return new (window.AudioContext || window.webkitAudioContext)(options)
  }
  catch (error) {
    throw new RuntimeError('AUDIO_CONTEXT_UNSUPPORTED', error)
  }
}

export function getOfflineAudioContext(options: OfflineAudioContextOptions) {
  try {
    return new OfflineAudioContext(options)
  }
  catch {
    return getAudioContext(options)
  }
}

/** Converts Blob to ArrayBuffer */
export function blobToBuffer(blob: Blob): Promise<ArrayBuffer> {
  if (blob instanceof Blob && blob.arrayBuffer)
    return blob.arrayBuffer()

  return new Promise<ArrayBuffer>((resolve, reject) => {
    const reader = new FileReader()

    reader.onload = () => resolve(reader.result as ArrayBuffer)
    reader.onerror = reject
    reader.readAsArrayBuffer(blob)
  })
}

/** Converts Blob or ArrayBuffer to AudioBuffer using a OfflineAudioContext if supported */
export async function getAudioBuffer(buffer: Blob | ArrayBuffer, config: OfflineAudioContextOptions = {
  numberOfChannels: 1,
  sampleRate: 48000,
  length: 48000
}): Promise<AudioBuffer> {
  const input = buffer instanceof Blob
    ? await blobToBuffer(buffer)

    // Clone buffer to prevent mutation
    : buffer.slice(0)

  return new Promise((resolve, reject) =>
    getOfflineAudioContext(config).decodeAudioData(input, resolve, reject)
  )
}

export function playBeep({
  frequency = 440,
  rampValue = 0.00001,
  rampDuration = 1
} = {}) {
  const context = getAudioContext({ latencyHint: 'playback' })
  const currentTime = context.currentTime
  const osc = context.createOscillator()
  const gain = context.createGain()

  osc.connect(gain)
  gain.connect(context.destination)

  gain.gain.value = 0.4
  gain.gain.setValueAtTime(gain.gain.value, currentTime)
  gain.gain.exponentialRampToValueAtTime(rampValue, currentTime + rampDuration)

  return new Promise<void>((resolve) => {
    // Cleanup
    osc.onended = () => {
      gain.disconnect(context.destination)
      osc.disconnect(gain)
      context.close()

      resolve()
    }

    osc.type = 'sine'
    osc.frequency.value = frequency
    osc.start(currentTime)
    osc.stop(currentTime + rampDuration)
  })
}
