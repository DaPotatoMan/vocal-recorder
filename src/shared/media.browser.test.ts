import audioUrl from '~/tests/assets/audio-sample.mp3?url'
import { RecorderError } from '..'
import { blobToBuffer, getAudioBuffer, getOfflineAudioContext, StreamUtil } from './media'

describe('class: StreamUtil', () => {
  let stream: MediaStream

  it('can get media stream', async () => {
    stream = await StreamUtil.get()
    expect(stream).toBeInstanceOf(MediaStream)
  })

  it('can get settings from media stream', () => {
    const settings = StreamUtil.getSettings(stream)

    expect(settings.deviceId).not.toBeFalsy()
    expect(settings.groupId).not.toBeFalsy()
    expect(settings.sampleRate).toBeGreaterThan(44100)
  })

  it('can validate active media stream', () => {
    const isValid = StreamUtil.isValid(stream)
    expect(isValid).toBe(true)
  })

  it('can dispose stream', () => {
    const isActive = StreamUtil.dispose(stream)
    expect(isActive).toBe(false)
    expect(StreamUtil.isValid(stream)).toBe(false)
  })

  it('should throw error when getUserMedia is unavailable', () => {
    const ref = navigator.mediaDevices.getUserMedia
    navigator.mediaDevices.getUserMedia = undefined as any

    expect(StreamUtil.get).rejects.toThrowError(RecorderError.NO_GETUSERMEDIA)

    // Restore
    navigator.mediaDevices.getUserMedia = ref
  })
})

describe('getOfflineAudioContext', () => {
  const octx = globalThis.OfflineAudioContext
  const ctx = globalThis.AudioContext
  const getContext = () => getOfflineAudioContext({ length: 48000, sampleRate: 48000, numberOfChannels: 1 })

  afterEach(() => {
    globalThis.OfflineAudioContext = octx
    globalThis.AudioContext = ctx
  })

  it('should create OfflineAudioContext', () => {
    expect(getContext()).toBeInstanceOf(OfflineAudioContext)
  })

  it('should use fallback AudioContext when OfflineAudioContext is not defined', () => {
    // @ts-expect-error - Testing fallback
    globalThis.OfflineAudioContext = undefined

    expect(getContext()).toBeInstanceOf(AudioContext)
  })

  it('should throw error when AudioContext is unavailable', () => {
    // @ts-expect-error - Testing fallback
    globalThis.OfflineAudioContext = undefined
    // @ts-expect-error - Testing fallback
    globalThis.AudioContext = undefined

    expect(getContext).toThrowError(RecorderError.NO_AUDIO_CONTEXT)
  })
})

describe('blobToBuffer', async () => {
  const blob = await fetch(audioUrl).then(res => res.blob())

  it('should convert Blob to ArrayBuffer', () => {
    expect(blobToBuffer(blob)).resolves.toBeInstanceOf(ArrayBuffer)
  })

  it('shoulw use FileReader if blob.arrayBuffer() not supported', () => {
    blob.arrayBuffer = undefined as any
    expect(blobToBuffer(blob)).resolves.toBeInstanceOf(ArrayBuffer)
  })

  it('should throw error for unknown type', () => {
    expect(() => blobToBuffer(null as any)).rejects.toThrow()
  })
})

describe('getAudioBuffer', async () => {
  const audio = await fetch(audioUrl)
  const audioBuffer = await audio.arrayBuffer()
  const audioBlob = new Blob([audioBuffer], { type: 'audio/mpeg' })

  function assertBuffer(buffer: AudioBuffer) {
    expect(buffer).toBeInstanceOf(AudioBuffer)
    expect(buffer.duration).toBeGreaterThan(23)
    expect(buffer.sampleRate).toBe(48000)
  }

  it('can parse Blob', async () => {
    const buffer = await getAudioBuffer(audioBlob)
    assertBuffer(buffer)

    // Ensure original blob is not mutated
    expect(audioBlob.size).toBeGreaterThan(0)
  })

  it('can parse ArrayBuffer', async () => {
    const buffer = await getAudioBuffer(audioBuffer)
    assertBuffer(buffer)

    // Ensure original buffer is not mutated
    expect(audioBuffer.byteLength).toBeGreaterThan(0)
  })
})
