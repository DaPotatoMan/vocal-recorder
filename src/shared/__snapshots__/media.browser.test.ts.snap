// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: StreamUtil > can get settings from media stream 1`] = `
[
  "autoGainControl",
  "channelCount",
  "deviceId",
  "echoCancellation",
  "groupId",
  "latency",
  "noiseSuppression",
  "sampleRate",
  "sampleSize",
  "voiceIsolation",
]
`;

exports[`class: StreamUtil > can get settings from media stream 2`] = `
[
  "autoGainControl",
  "channelCount",
  "deviceId",
  "echoCancellation",
  "groupId",
  "latency",
  "noiseSuppression",
  "sampleRate",
  "sampleSize",
  "voiceIsolation",
]
`;
